reference/reference.go

```diff
@@ -195,40 +204,139 @@ func ParseNamed(s string) (Named, error) {
 	return named, nil
 }
 
-// WithName returns a named object representing the given string. If the input
-// is invalid ErrReferenceInvalidFormat will be returned.
-func WithName(name string) (Named, error) {
+// NamedOnly returns true if reference only contains a repo name and not
+// other modifiers.
+func NamedOnly(ref Named) bool {
+	switch ref.(type) {
+	case Tagged:
+		return false
+	case Canonical:
+		return false
+	case Digested:
+		return false
+	}
+
+	return true
+}
+
+// NameOnly drops other reference information and only retains the name.
+func NameOnly(ref Named) Named {
+	return repository(ref.Name())
+}
+
+// WithName replaces the Name() on the reference with the provided value of
+// name.
+//
+// If the target of named is from another package, any data on the backing
+// type will be lost and replaced with an instance from this package. Simply
+// put, the resulting type will only include information available on the
+// interface.
+func WithName(ref Reference, name string) (Named, error) {
+	if name == "" {
+		return nil, ErrNameEmpty
+	}
+
 	if len(name) > NameTotalLengthMax {
 		return nil, ErrNameTooLong
 	}
+
 	if !anchoredNameRegexp.MatchString(name) {
 		return nil, ErrReferenceInvalidFormat
 	}
-	return repository(name), nil
+
+	if ref == nil {
+		return repository(name), nil
+	}
+
+	switch v := ref.(type) {
+	case repository:
+		return v, nil
+	case reference:
+		v.name = name
+		return v, nil
+	case digestReference:
+		return canonicalReference{
+			name:   name,
+			digest: v.Digest(),
+		}, nil
+	case taggedReference:
+		v.name = name
+		return v, nil
+	case canonicalReference:
+		v.name = name
+		return v, nil
+	}
+
+	return nil, ErrNameDisallowed
 }
 
 // WithTag combines the name from "name" and the tag from "tag" to form a
-// reference incorporating both the name and the tag.
-func WithTag(name Named, tag string) (NamedTagged, error) {
+// reference incorporating both the name and the tag. When the reference
+// cannot be tagged, ErrCannotTagReference is returned.
+//
+// If the target of named is from another package, any data on the backing
+// type will be lost and replaced with an instance from this package. Simply
+// put, the resulting type will only include information available on the
+// interface.
```

Comment:
> Can we go back to the original implementations of WithTag and WithDigest?

I fixed the implementation. Really, we need to restrict the types that can implement `reference` for these combinators to work correctly.